firmware/development/logic_analyzer.cpp

```diff
@@ -88,44 +82,42 @@ void WaveReader::onFallEvent() {
 	periodEventTimeUs = nowUs;
 }
 
-static void waIcuPeriodCallback(WaveReader *reader) {
-	reader->onFallEvent();
+void logicAnalyzerCallback(void* arg, efitick_t stamp) {
+	WaveReader* instance = reinterpret_cast<WaveReader*>(arg);
+
+	bool rise = palReadLine(instance->line) == PAL_HIGH;
+
+	if(rise) {
+		riseCallback(instance);
+	} else {
+		instance->onFallEvent();
+	}
 }
 
 static void initWave(const char *name, int index) {
 	brain_pin_e brainPin = engineConfiguration->logicAnalyzerPins[index];
 
-	waveReaderCount++;
-	efiAssertVoid(CUSTOM_ERR_6655, index < MAX_ICU_COUNT, "too many ICUs");
+	efiAssertVoid(CUSTOM_ERR_6655, index < efi::size(readers), "too many ICUs");
 	WaveReader *reader = &readers[index];
 
 	if (!isBrainPinValid(brainPin)) {
 		/**
 		 *  in case we are running, and we select none for a channel that was running, 
 		 *  this way we ensure that we do not get false report from that channel 
 		 **/
-		reader->hw = nullptr;
+		reader->line = 0;
 		return;
 	}
-		
 
 	reader->name = name;
 
-	reader->hw = startDigitalCapture("wave input", brainPin);
+	reader->line = PAL_LINE(getHwPort("logic", brainPin), getHwPin("logic", brainPin));
```

Comment:
c'mon

the smart way would have been to use the suggested universal input caputure API

Reply:
....an api that doesn't exist.